hazelcast/src/main/java/com/hazelcast/sql/impl/expression/math/RoundTruncateFunction.java

```diff
@@ -0,0 +1,219 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.math;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpressionWithType;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+/**
+ * Implementation of ROUND/TRUNCATE functions.
+ */
+public class RoundTruncateFunction<T> extends BiExpressionWithType<T> implements IdentifiedDataSerializable {
+
+    private boolean truncate;
+
+    public RoundTruncateFunction() {
+        // No-op.
+    }
+
+    private RoundTruncateFunction(Expression<?> operand1, Expression<?> operand2, QueryDataType resultType, boolean truncate) {
+        super(operand1, operand2, resultType);
+
+        this.truncate = truncate;
+    }
+
+    public static Expression<?> create(
+        Expression<?> operand1,
+        Expression<?> operand2,
+        QueryDataType resultType,
+        boolean truncate
+    ) {
+        if (operand2 == null) {
+            QueryDataType operand1Type = operand1.getType();
+
+            // No conversion is expected for non-fractional types when the length operand is not defined.
+            if (MathFunctionUtils.notFractional(operand1Type)) {
+                assert operand1Type == resultType;
+
+                return operand1;
+            }
+        }
+
+        return new RoundTruncateFunction<>(operand1, operand2, resultType, truncate);
+    }
+
+    @SuppressWarnings({"unchecked", "checkstyle:CyclomaticComplexity", "checkstyle:ReturnCount"})
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        // Get base operand.
+        Object operand1Value = operand1.eval(row, context);
+
+        // NULL always yields NULL
+        if (operand1Value == null) {
+            return null;
+        }
+
+        // Get length.
+        Integer operand2Value = operand2 != null ? MathFunctionUtils.asInt(operand2, row, context) : null;
```

Comment:
This line may produce a cryptic conversion error since operand type checker for `ROUND` and `TRUNCATE` is defined as `NUMERIC_OPTIONAL_INTEGER`. That last `INTEGER` is referring to `SqlTypeFamily.INTEGER` (not to `SqlTypeName.INTEGER`) and that family refers to all integer types including `BIGINT`. Either the type checker should allow only the compatible types or some meaningful error message should be provided.

Reply:
Fixed and tested in https://github.com/hazelcast/hazelcast/pull/17340/commits/5ce9f81ee0c5574f58c1197f9107bea75f3c9e51